src/rust/engine/src/fs.rs

```diff
@@ -635,27 +636,77 @@ impl fmt::Debug for Snapshot {
   }
 }
 
+// Like std::fs::create_dir_all, except handles concurrent calls among multiple
+// threads or processes. Originally lifted from rustc.
+fn safe_create_dir_all_ioerror(path: &Path) -> Result<(), io::Error> {
+  match fs::create_dir(path) {
+    Ok(()) => return Ok(()),
+    Err(ref e) if e.kind() == io::ErrorKind::AlreadyExists => return Ok(()),
+    Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
+    Err(e) => return Err(e),
+  }
+  match path.parent() {
+    Some(p) => try!(safe_create_dir_all_ioerror(p)),
+    None => return Ok(()),
+  }
+  match fs::create_dir(path) {
+    Ok(()) => Ok(()),
+    Err(ref e) if e.kind() == io::ErrorKind::AlreadyExists => Ok(()),
+    Err(e) => Err(e),
+  }
+}
+
+fn safe_create_dir_all(path: &Path) -> Result<(), String> {
+  safe_create_dir_all_ioerror(path).map_err(|e| format!("Failed to create dir {:?} due to {:?}", path, e))
+}
+
+fn safe_create_tmpdir_in(base_dir: &Path, prefix: &str) -> Result<TempDir, String> {
+  safe_create_dir_all(&base_dir)?;
+  Ok(
+    TempDir::new_in(&base_dir, prefix)
+      .map_err(|e| format!("Failed to create tempdir {:?} due to {:?}", base_dir, e))?
+  )
+}
+
 /**
- * A facade for the snapshot directory, which is currently thrown away at the end of each run.
+ * A facade for the snapshot directory, which lives under the pants workdir.
  */
 pub struct Snapshots {
-  temp_dir: TempDir,
-  next_temp_id: atomic::AtomicUsize,
+  snapshots_dir: PathBuf,
+  snapshots_generator: Arc<Mutex<(TempDir, usize)>>,
```

Comment:
This doesn't need to be wrapped in an `Arc`, as far as I can tell? You'd need an `Arc` if this value needed to be `Send` ie, sendable between threads. Try without the `Arc`.